File: frontend/lib/models/app_notification.dart
/// Mirrors `NotificationResource` from the backend.
///
/// Named AppNotification to avoid shadowing dart:html's Notification.
class AppNotification {
  const AppNotification({
    required this.id,
    required this.title,
    required this.message,
    required this.type,
    this.loanId,
    required this.isRead,
    this.createdAt,
  });

  final int id;
  final String title;
  final String message;

  /// One of: loan_created | loan_approved | loan_rejected |
  ///          loan_borrowed | loan_returned | system
  final String type;

  final int? loanId;
  final bool isRead;
  final DateTime? createdAt;

  // ---- Type constants (match backend) ----
  static const String typeLoanCreated = 'loan_created';
  static const String typeLoanApproved = 'loan_approved';
  static const String typeLoanRejected = 'loan_rejected';
  static const String typeLoanBorrowed = 'loan_borrowed';
  static const String typeLoanReturned = 'loan_returned';
  static const String typeSystem = 'system';

  factory AppNotification.fromJson(Map<String, dynamic> json) {
    return AppNotification(
      id: json['id'] as int,
      title: json['title'] as String,
      message: json['message'] as String,
      type: json['type'] as String? ?? typeSystem,
      loanId: json['loan_id'] as int?,
      isRead: json['is_read'] as bool? ?? false,
      createdAt: _parseDate(json['created_at']),
    );
  }

  AppNotification copyWith({bool? isRead}) {
    return AppNotification(
      id: id,
      title: title,
      message: message,
      type: type,
      loanId: loanId,
      isRead: isRead ?? this.isRead,
      createdAt: createdAt,
    );
  }

  static DateTime? _parseDate(Object? raw) {
    if (raw is String && raw.isNotEmpty) return DateTime.tryParse(raw);
    return null;
  }
}

File: backend/app/Models/AppNotification.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

/**
 * In-app notification record.
 *
 * Named AppNotification (not Notification) to avoid clashing with
 * Laravel's own \Illuminate\Notifications\Notification base class.
 *
 * @property int $id
 * @property int $user_id
 * @property string $title
 * @property string $message
 * @property string $type  one of the TYPE_* constants
 * @property int|null $loan_id
 * @property bool $is_read
 * @property \Illuminate\Support\Carbon $created_at
 * @property \Illuminate\Support\Carbon $updated_at
 */
class AppNotification extends Model
{
    protected $table = 'notifications';

    // ---- Type constants ----
    public const TYPE_LOAN_CREATED  = 'loan_created';
    public const TYPE_LOAN_APPROVED = 'loan_approved';
    public const TYPE_LOAN_REJECTED = 'loan_rejected';
    public const TYPE_LOAN_BORROWED = 'loan_borrowed';
    public const TYPE_LOAN_RETURNED = 'loan_returned';
    public const TYPE_SYSTEM        = 'system';

    /** @var list<string> */
    protected $fillable = [
        'user_id',
        'title',
        'message',
        'type',
        'loan_id',
        'is_read',
    ];

    /**
     * @return array<string, string>
     */
    protected function casts(): array
    {
        return [
            'is_read' => 'boolean',
        ];
    }

    // ---- Relationships ----

    /**
     * @return BelongsTo<User, $this>
     */
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /**
     * @return BelongsTo<Loan, $this>
     */
    public function loan(): BelongsTo
    {
        return $this->belongsTo(Loan::class);
    }
}
